io.openems.edge.controller.symmetric.selfconsumption/src/io/openems/edge/controller/symmetric/selfconsumption/SelfConsumption.java

```diff
@@ -0,0 +1,117 @@
+import io.openems.edge.controller.api.Controller;
+import io.openems.edge.ess.api.ManagedSymmetricEss;
+import io.openems.edge.ess.power.api.Phase;
+import io.openems.edge.ess.power.api.Power;
+import io.openems.edge.ess.power.api.Pwr;
+import io.openems.edge.ess.power.api.Relationship;
+import io.openems.edge.meter.api.SymmetricMeter;
+
+@Designate(ocd = Config.class, factory = true)
+@Component(name = "Controller.Symmetric.SelfConsmption", immediate = true, configurationPolicy = ConfigurationPolicy.REQUIRE)
+public class SelfConsumption extends AbstractOpenemsComponent implements Controller, OpenemsComponent {
+
+	@Reference
+	protected ComponentManager componentManager;
+
+	private int gridPower;
+
+	private final Logger log = LoggerFactory.getLogger(SelfConsumption.class);
+
+	private Config config = null;
+
+	public enum ChannelId implements io.openems.edge.common.channel.ChannelId {
+		GRID_POWER(Doc.of(OpenemsType.INTEGER) //
+				.text("Actual Grid Power")); //
+
+		private final Doc doc;
+
+		private ChannelId(Doc doc) {
+			this.doc = doc;
+		}
+
+		@Override
+		public Doc doc() {
+			return this.doc;
+		}
+	}
+
+	public SelfConsumption() {
+		super(//
+				OpenemsComponent.ChannelId.values(), //
+				Controller.ChannelId.values(), //
+				ChannelId.values());
+	}
+
+	@Activate
+	void activate(ComponentContext context, Config config) {
+		super.activate(context, config.id(), config.alias(), config.enabled());
+		this.config = config;
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		super.deactivate();
+	}
+
+	@Override
+	public void run() throws OpenemsNamedException {
+
+		// Get required variables
+		ManagedSymmetricEss ess = this.componentManager.getComponent(this.config.ess_id());
+		SymmetricMeter meter = this.componentManager.getComponent(this.config.meter_id());
+
+		/*
+		 * Check that we are On-Grid (and warn on undefined Grid-Mode)
+		 */
+		GridMode gridMode = ess.getGridMode().value().asEnum();
+		if (gridMode.isUndefined()) {
+			this.logWarn(this.log, "Grid-Mode is [UNDEFINED]");
+		}
+		if (gridMode != GridMode.ON_GRID) {
+			return;
+		}
+
+		// Calculating the actual grid power
+		this.gridPower = meter.getActivePower().value().orElse(0); /* current buy-from/sell-to grid */
+
+		// Setting the Target hour channel id
+		IntegerReadChannel actualGridPower = this.channel(ChannelId.GRID_POWER);
+		actualGridPower.setNextValue(this.gridPower);
+
+		// Checking if the grid power is above the maximum feed-in
+		if ((this.gridPower * -1) > this.config.Maximum_Feed_In()) {
+
+			// Adjusting the limit to the maximum feed-in.
+			int calculatedPower = this.gridPower + this.config.Maximum_Feed_In();
+
+			Power power = ess.getPower();
+			calculatedPower = power.fitValueIntoMinMaxPower(this.id(), ess, Phase.ALL, Pwr.ACTIVE, calculatedPower);
+
+			/*
+			 * set result
+			 */
+			ess.addPowerConstraintAndValidate("SelfConsmption-OptimizationController", Phase.ALL, Pwr.ACTIVE,
+					Relationship.EQUALS, calculatedPower); //
```

Comment:
I don't think this is the best way to design this controller. I would expect you to only give a GREATER_OR_EQUALS contraint and letting the normal 'Balancing'-Controller do the rest. Then this controller would not be a 'SelfConsumption' controller but a 'AvoidMaximumFeedIn'-Controller. Otherwise you need to document why you need this controller additionally to the existing Balancing-Controller,

i.e. how does Service know, when they need to activate this Controller and when to activate traditional Balancing-Controller?

Reply:
well its my mistake. I had two repositories when i was working on this. one with KACO implementation(fems) and one with normal openems one. i was testing and making changes on one but didn't do this change in this repository and pushed it without checking.